### ISSUE TITLE
Question - about the credit_window and the size of the circular buffer

### ISSUE THREAD

**Issue by amarzavery** [amqp/rhea#73]
- Is there any relationship between the credit_window and the size of the circular buffer?
- What happens when the credit window is set to 3000 a value greater than the current fixed size of the circular buffer (2048)?
- If `autoaccept` on the receiver link is set to `false` and the `credit_window` is set to `3000` (a value greater than 2048) and the client receives 3000 messages. The client is processing those messages and for some reason has not `completed|abondon|rejected|released|modified` them. Then during that time will the client be able to receive more messages from the service?  How would the circular buffer look like at that time?
- What would happen in the above scenario when `autoaccept` is set to `false` and the `credit_window` is set to `0`. So how would the message flow look like?  (I assume it will be dependent on the number of credits the client will issue on the receiver link). How would the circular buffer look like in this scenario?

**Comment by grs** [comment]
> Is there any relationship between the credit_window and the size of the circular buffer?

At present there is no relationship

> What happens when the credit window is set to 3000 a value greater than the current size of the circular buffer (2048)?

The circular buffer sets an implicit limit on the number of unsettled messages that can be held. There should be some more explicit checking enforcing.

> If autoaccept on the receiver link is set to false and the credit_window is set to 3000 (a value greater than 2048) and the client receives 3000 messages. The client is processing those messages and for some reason has not completed|abondon|rejected|released|modified them. Then during that time will the client be able to receive more messages from the service? How would the circular buffer look like at that time?

You would get a buffer overflow error in this case

> What would happen in the above scenario when autoaccept is set to false and the credit_window is set to 0. So how would the message flow look like? (I assume it will be dependent on the number of credits the client will issue on the receiver link). How would the circular buffer look like in this scenario?

If you allow more unsettled deliveries to arrive than the circular buffer can handle, it will raise an overflow error.

**Comment by amarzavery** [comment]
Thank you for answering the above questions. 

I think the size of the circular buffer should be allowed to be configured in atleast two ways:

1. Establish an explicit relationship between the credit_window and the circular buffer. For instance the size of the circular buffer should be a minimum of 2048 or 100 + the size of credit_window if the credit_window is greater than 2048. If possible let the customer select what that delta should be (100 or something else).
2. Let the customer specify the size of the circular buffer. This will always override what happens in step 1.

What do you think?

**Comment by grs** [comment]
I agree that allowing it to be directly configured (i.e. option 2.) makes sense. It is also true that the outstanding credit should never be greater that the remaining capacity in the incoming buffer; that I think would be better handled by a check when issuing credit.
